### PR TITLE
MVP CLI - draft 0 

### PR DESCRIPTION
**Pull request by jayden-sudo**

### CODE REVIEW

File: wallet-core-travel/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.DS_Store
```
Comment (jhfnetboy) [inline wallet-core-travel/.gitignore:1]: 这个是图片的cache缓存吧，有空可以清理，没空就不管

File: wallet-core-travel/package.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "scripts": {
+    "start": "ts-node src/app.ts",
+    "build": "tsc && node dist/app.js"
+  },
+  "devDependencies": {
```
Comment (jhfnetboy) [inline wallet-core-travel/package.json:6]: So we can define all the developers' client dev environment: Node ? ts-node? ts?

Comment (jhfnetboy) [reply wallet-core-travel/package.json:6]: we can set a agreement on version number

File: wallet-core-travel/src/ABI/entry_point.json [hunk 3]
```diff
@@ -0,0 +1,856 @@
+[
```
Comment (jhfnetboy) [inline wallet-core-travel/src/ABI/entry_point.json:1]: So where is this ABI's original contract? come from this URL? https://github.com/proofofsoulprotocol/4337-MVP/tree/develop/contracts

File: wallet-core-travel/src/app.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/*
+ * @Description: 
+ * @Version: 1.0
+ * @Autor: z.cejay@gmail.com
+ * @Date: 2022-06-22 16:15:42
+ * @LastEditors: cejay
+ * @LastEditTime: 2022-07-26 10:52:31
+ */
+import Web3 from 'web3';
+import { Utils } from './utils/utils';
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:10]: @earthbeyond  so how about our directory archi ? the CLI mode wallet can transfer into the Crx version wallet

File: wallet-core-travel/.gitignore (hunk 1 above)
Comment (jayden-sudo) [reply wallet-core-travel/.gitignore:1]: 这个是特意加上的，macos 不加的话 就会多提交一些东西

File: wallet-core-travel/src/app.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import fs from 'fs';
+import { UserOperation } from './entity/userOperation';
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:12]: userOperation.js we can have a fast talk to share the core dev info to the team

File: wallet-core-travel/src/app.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import { EntryPointAddress, Create2Factory, AddressZero } from './defines';
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:13]: So the defines directory is the uneditable vendor resource?

File: wallet-core-travel/src/app.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import { signUserOp, getRequestId } from './utils/UserOp';
+
+async function main() {
+
+    /**
+     * web3 instance
+     */
+    const web3 = new Web3('https://ropsten.infura.io/v3/21d2f7b2f3cd4afca4c6d1930ec801c4');
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:21]: we use infura as the outer ETH relay? we can apply grant for this^_^ @zengjiajun0623

File: wallet-core-travel/src/app.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    // get chainId ( use for generate signature)
+    const chainId = await web3.eth.net.getId();
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:24]: It will adapt to Layer2?

File: wallet-core-travel/src/app.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    console.log(`chainId: ${chainId}`);
+
+    /**
+     * contract bytecode.contract template (test code)
+     */
+    const bytecodeTemplate = fs.readFileSync('./src/bytecodes/simpleWallet.txt', 'utf8');
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:30]: It will be a version hash in IPFS or some trustless storage?

File: wallet-core-travel/src/app.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+    // generate control key for EIP-4337 contract
+    const _account = web3.eth.accounts.create();
+    const address = _account.address;
+    const privateKey = _account.privateKey;
+    console.log(`Address: ${address}, PrivateKey: ${privateKey}`);
+
+    // generate contract wallet address
+    // create bytecode with constructor params(entry point address, control key[address])
+    const initCode = `${bytecodeTemplate}000000000000000000000000${EntryPointAddress.substring(2).toLowerCase()}000000000000000000000000${address.substring(2).toLowerCase()}`;
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:40]: It was a fixed method to get initCode? we can set it to Utils? or only here use this?

File: wallet-core-travel/src/app.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        userOperation.sender = walletAddress;
+        userOperation.nonce = 0;
+        userOperation.initCode = initCode;
+        //userOperation.callData = '0x';
+        userOperation.callGas = 1e6;
+        userOperation.verificationGas = 1e6;
+        userOperation.preVerificationGas = 1e6;
+        userOperation.maxFeePerGas = 10e9;
+        userOperation.maxPriorityFeePerGas = 10e9;
+        userOperation.paymaster = AddressZero; // paymaster is not used in this example
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:63]: How about Addzero? hh

File: wallet-core-travel/src/app.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    // 2. entrypoint::simulateValidation(UserOperation)
+    // read entrypoint ABI from file
+    const entrypointAbi = JSON.parse(fs.readFileSync('./src/ABI/entry_point.json', 'utf8'));
+    // get entrypoint contract instance
+    const entrypointContract = new web3.eth.Contract(entrypointAbi, EntryPointAddress);
+    // call entrypoint::simulateValidation(UserOperation)
+    // function simulateValidation(UserOperation calldata userOp) external returns (uint256 preOpGas, uint256 prefund)
+    try {
+        const result = await entrypointContract.methods.simulateValidation(userOperation).call();
```
Comment (jhfnetboy) [inline wallet-core-travel/src/app.ts:78]: So simulation returns what?

File: wallet-core-travel/src/defines.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+ * @LastEditTime: 2022-07-26 10:13:02
+ */
+
+
+export const AddressZero = '0x0000000000000000000000000000000000000000';
+
+/**
+ * entry point contract address. singleton contract.(test address)
+ */
+export const EntryPointAddress = '0xeC97dc3Ef34b876caA9CC3E7EF91DE0D66cd689F';
```
Comment (jhfnetboy) [inline wallet-core-travel/src/defines.ts:16]: if set to .env, we change the address and needn't to compile the ts script?

File: wallet-core-travel/src/defines.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * create2factory address defined in eip-2470.
+ */
+export const Create2Factory = '0xce0042B868300000d44A59004Da54A005ffdcf9f';
```
Comment (jhfnetboy) [inline wallet-core-travel/src/defines.ts:21]: who set this factory addr? EF or us ?

File: wallet-core-travel/src/utils/UserOp.ts [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        { type: 'uint256', name: 'maxFeePerGas' },
+        { type: 'uint256', name: 'maxPriorityFeePerGas' },
+        { type: 'address', name: 'paymaster' },
+        { type: 'bytes', name: 'paymasterData' },
+        { type: 'bytes', name: 'signature' }
+      ],
+      name: 'userOp',
+      type: 'tuple'
+    }
+    let encoded = defaultAbiCoder.encode([userOpType as any], [{ ...op, signature: '0x' }])
```
Comment (jhfnetboy) [inline wallet-core-travel/src/utils/UserOp.ts:39]: I need some background to know this sentence....
for ex, .encode is a ENUM list? how can we select a special algorithm?

Comment (jhfnetboy) [reply wallet-core-travel/src/utils/UserOp.ts:39]: and how about the signature be produced?

File: wallet-core-travel/src/utils/UserOp.ts [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    op.preVerificationGas,
+    op.maxFeePerGas,
+    op.maxPriorityFeePerGas,
+    op.paymaster,
+    keccak256(op.paymasterData)
+  ])
+}
+
+export function getRequestId(op: UserOperation, chainId: number): string {
+  const userOpHash = keccak256(packUserOp(op, true))
```
Comment (jhfnetboy) [inline wallet-core-travel/src/utils/UserOp.ts:93]: the hash will be produced by keccak256? it was fixed?

**Review by jhfnetboy — COMMENTED** [review]
sorry for so many questions, we can have a review talk about this:)

**Pull request merged**